### PR TITLE
feat: Add GaleShapley in a new folder Greedy

### PR DESCRIPTION
**Pull request by BKarthik7**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [x] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.

### CODE REVIEW

**Comment by BKarthik7** [comment]
@appgurueu @raklaptudirm can you please review

**Comment by BKarthik7** [comment]
@raklaptudirm can you please review

File: Greedy/test/GaleShapley.test.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { expect, test } from 'vitest'
+import { stableMatching } from '../GaleShapley'
+
+test('Test Case 1', () => {
```
Comment (appgurueu) [inline Greedy/test/GaleShapley.test.js:4]: A test case structure like this isn't helpful. Please see https://github.com/TheAlgorithms/TypeScript/blob/master/CONTRIBUTING.md#writing-good-tests (it's for the TS repo, but translates to JS just as well).

TL;DR: Use a `describe` block for Gale-Shapley and `it.each` for individual test cases if you have no descriptive labels for them.

File: Greedy/GaleShapley.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * @param {number[][]} recipientPref Preferences of recipients, where each recipient has an ordered list of donors.
+ * @returns {number[]} Array where the index is the donor, and the value at the index is the matched recipient.
+ *
+ * @example
+ * const donorPref = [[0, 1, 3, 2], [0, 2, 3, 1], [1, 0, 2, 3], [0, 3, 1, 2]];
+ * const recipientPref = [[3, 1, 2, 0], [3, 1, 0, 2], [0, 3, 1, 2], [1, 0, 3, 2]];
+ * stableMatching(donorPref, recipientPref); // Output: [1, 2, 3, 0]
+ */
+function stableMatching(donorPref, recipientPref) {
+  // Initialize the number of donors and create a list of unmatched donors
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:29]: I'd use this array as a stack because that's efficient.

File: Greedy/GaleShapley.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  let n = donorPref.length
+  let unmatchedDonors = Array.from({ length: n }, (_, i) => i)
+
+  // Records of which recipient each donor is paired with and vice versa
+  let donorRecord = Array(n).fill(-1) // Donor to recipient
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:34]: I'd compute `donorRecord` at the end (or not at all even and instead just return `recRecord` and document that because the two are equivalent). That way you save yourself having to remember to update it during the algorithm, and the constant factor of the $O(n^2)$ many operations is better.

File: Greedy/GaleShapley.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  let recRecord = Array(n).fill(-1) // Recipient to donor
+
+  // Array to track how many recipients each donor has proposed to
+  let numDonations = Array(n).fill(0)
+
+  // While there are unmatched donors
+  while (unmatchedDonors.length > 0) {
+    // Take the first unmatched donor
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:42]: Don't. Take the last one instead; use the array as a stack (you could also use a queue, but why do that when a stack suffices). If you take the first one, you have to do a costly $O(n)$ unshift / splice later on to remove it (arguably you could also swap with the last one and pop, but that's just unnecessary complexity when you can just pop here), which wrecks your time complexity: You get $O(n^3)$ instead of $O(n^2)$.

File: Greedy/GaleShapley.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    let donor = unmatchedDonors[0]
+    let donorPreference = donorPref[donor]
+
+    // Find the next recipient this donor prefers
+    let recipient = donorPreference[numDonations[donor]]
+    numDonations[donor] += 1
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:48]:
```suggestion
    numDonations[donor]++
```

File: Greedy/GaleShapley.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * stableMatching(donorPref, recipientPref); // Output: [1, 2, 3, 0]
+ */
+function stableMatching(donorPref, recipientPref) {
+  // Initialize the number of donors and create a list of unmatched donors
+  let n = donorPref.length
+  let unmatchedDonors = Array.from({ length: n }, (_, i) => i)
+
+  // Records of which recipient each donor is paired with and vice versa
+  let donorRecord = Array(n).fill(-1) // Donor to recipient
+  let recRecord = Array(n).fill(-1) // Recipient to donor
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:35]: I also think it'd be idiomatic to use `null` rather than `-1` to signal the absence of a value.

File: Greedy/GaleShapley.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    let recipient = donorPreference[numDonations[donor]]
+    numDonations[donor] += 1
+
+    // Get recipient's preference list and check the current match
+    let recPreference = recipientPref[recipient]
+    let prevDonor = recRecord[recipient]
+
+    // If recipient is already matched, check if they prefer the new donor
+    if (prevDonor !== -1) {
+      if (recPreference.indexOf(prevDonor) > recPreference.indexOf(donor)) {
```
Comment (appgurueu) [inline Greedy/GaleShapley.js:56]: This also wrecks time complexity by adding a linear factor. Instead you should precompute the inverse mappings initially as lookup tables (this is what's responsible for the $O(n^2)$ best case): For each receiver, compute the mapping from donor -> preference ahead of time.

**Review by appgurueu — CHANGES_REQUESTED** [review]
